pkg/queue/breaker_test.go

```diff
@@ -121,43 +120,56 @@ func TestBreakerLargeCapacityRecover(t *testing.T) {
 
 // Test empty semaphore, token cannot be acquired
 func TestSemaphore_Get_HasNoCapacity(t *testing.T) {
-	want := int32(0)
-	acquired := int32(0)
+	gotChan := make(chan struct{}, 1)
+
 	sem := NewSemaphore(1, 0)
-	tryAcquire(sem, &acquired, 0)
+	tryAcquire(sem, gotChan)
 
-	// wait in case `acquired` changes
-	time.Sleep(semSleepInterval)
-	assertEqual(want, atomic.LoadInt32(&acquired), t)
+	select {
+	case <-gotChan:
+		t.Error("Token was acquired but shouldn't have been")
+	case <-time.After(20 * time.Millisecond):
+		// Test succeeds, semaphore didn't change in configured time
+	}
 }
 
 // Test empty semaphore, add capacity, token can be acquired
 func TestSemaphore_Get_HasCapacity(t *testing.T) {
-	want := int32(1)
-	acquired := int32(0)
+	gotChan := make(chan struct{}, 1)
+
 	sem := NewSemaphore(1, 0)
-	tryAcquire(sem, &acquired, 0)
+	tryAcquire(sem, gotChan)
 	sem.Release()
 
-	// to allow `acquired` to change
-	time.Sleep(semSleepInterval)
-	assertEqual(want, atomic.LoadInt32(&acquired), t)
+	select {
+	case <-gotChan:
+		// Test successful, successfully acquired a token.
+	case <-time.After(semAcquireTimeout):
+		t.Errorf("Was not able to acquire token before timeout")
+	}
 }
 
 //Test all put items can be consumed
 func TestSemaphore_Put(t *testing.T) {
-	want := int32(2)
+	gotChan := make(chan struct{}, 1)
+	wantAcquires := 2
+
 	requests := 3
-	var acquired int32
 	sem := NewSemaphore(2, 0)
 	for i := 0; i < requests; i++ {
-		tryAcquire(sem, &acquired, i)
+		tryAcquire(sem, gotChan)
 	}
 	sem.Release()
 	sem.Release()
 
-	time.Sleep(semSleepInterval)
-	assertEqual(want, atomic.LoadInt32(&acquired), t)
+	for i := 0; i < wantAcquires; i++ {
+		select {
```

Comment:
maybe abstract this 'select' into a method `assertTokenAquired(gotChan chan struct{})` or `assertTokenAquired(do func(), gotChan chan struct{})` if we want to include HasNoCapacity test as well?

Reply:
Will do, I'll leave the "HasNoCapacityTest" out though, seems to be custom enough to leave it a single occurrence.